sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala

```diff
@@ -0,0 +1,195 @@
+        new Iterator[UnsafeRow] {
+          private[this] val unsafeRow = new UnsafeRow()
+          private[this] var currOffset: Long = 0
+          private[this] var _nextRowSize = getNextRowSize()
+          private[this] def getNextRowSize(): Int = {
+            if (currOffset >= block.size()) {
+              -1
+            } else {
+              Platform.getInt(block.getBaseObject, block.getBaseOffset + currOffset)
+            }
+          }
+          override def hasNext: Boolean = _nextRowSize != END_OF_BLOCK
+          override def next(): UnsafeRow = {
+            assert(_nextRowSize > 0)
+            currOffset += 4
+            unsafeRow.pointTo(
+              block.getBaseObject, block.getBaseOffset + currOffset, numFields, _nextRowSize)
+            currOffset += _nextRowSize
+            _nextRowSize = getNextRowSize()
+            unsafeRow
+          }
+        }
+      }
+    }
+  }
+
+  private def compressBlock(
+      memoryBlock: MemoryBlock,
+      compressionCodec: CompressionCodec,
+      taskMemoryManager: TaskMemoryManager): MemoryBlock = {
+    // Compress the block using an on-heap byte array
+    val compressedBlockArray: Array[Byte] = {
+      val blockArray = new Array[Byte](memoryBlock.size().toInt)
+      Platform.copyMemory(
+        memoryBlock.getBaseObject,
+        memoryBlock.getBaseOffset,
+        blockArray,
+        Platform.BYTE_ARRAY_OFFSET,
+        memoryBlock.size())
+      val baos = new ByteArrayOutputStream(memoryBlock.size().toInt)
+      val compressedBaos = compressionCodec.compressedOutputStream(baos)
+      compressedBaos.write(blockArray)
+      compressedBaos.flush()
+      compressedBaos.close()
+      baos.toByteArray
+    }
+
+    // Allocate a new block with compressed byte array padded to word boundary
+    val totalRecordSize = compressedBlockArray.length + 4 // data + int to store size of padding
+    val nearestWordBoundary =
+      ByteArrayMethods.roundNumberOfBytesToNearestWord(totalRecordSize)
+    val padding = nearestWordBoundary - totalRecordSize
+    val compressedBlock = taskMemoryManager.allocateUnchecked(totalRecordSize + padding)
+    Platform.putInt(
+      compressedBlock.getBaseObject,
+      compressedBlock.getBaseOffset,
+      padding)
+    Platform.copyMemory(
+      compressedBlockArray,
+      Platform.BYTE_ARRAY_OFFSET,
+      compressedBlock.getBaseObject,
+      compressedBlock.getBaseOffset + 4,
+      compressedBlockArray.length)
+    taskMemoryManager.freeUnchecked(memoryBlock)
+    compressedBlock
+  }
+
+  private def decompressBlock(
+      compressedMemoryBlock: MemoryBlock,
+      decompressedSize: Int,
+      compressionCodec: CompressionCodec): MemoryBlock = {
+    // Copy compressed block (excluding padding) to on-heap byte array
+    val padding =
+      Platform.getInt(compressedMemoryBlock.getBaseObject, compressedMemoryBlock.getBaseOffset)
+    val compressedBlockArray = new Array[Byte](compressedMemoryBlock.size().toInt - padding)
```

Comment:
I think @JoshRosen mentioned earlier that we don't need to reinitialize here and can reuse this `Array` across multiple blocks